### PR TITLE
[flutter_tools] split project parsing/validation into separate function

### PR DESCRIPTION
**Pull request by jonahwilliams**
## Description

In order to share logic across build/archive/ipa commands, move the project parsing/validation into a different function. 

https://github.com/flutter/flutter/pull/64683

### CODE REVIEW

File: packages/flutter_tools/test/general.shard/ios/ios_device_start_nonprebuilt_test.dart [hunk 1]
```diff
@@ -43,17 +43,16 @@ const List<String> kRunReleaseArgs = <String>[
   '-configuration',
   'Release',
   '-quiet',
```
Comment (jonahwilliams) [inline packages/flutter_tools/test/general.shard/ios/ios_device_start_nonprebuilt_test.dart:45]: Ordering shouldn't matter here

File: packages/flutter_tools/test/general.shard/ios/ios_device_start_nonprebuilt_test.dart [hunk 2]
```diff
@@ -43,17 +43,16 @@ const List<String> kRunReleaseArgs = <String>[
...
   '-workspace',
   'Runner.xcworkspace',
   '-scheme',
   'Runner',
   'BUILD_DIR=/build/ios',
-  '-sdk',
-  'iphoneos',
   'ONLY_ACTIVE_ARCH=YES',
   'ARCHS=arm64',
-  'FLUTTER_SUPPRESS_ANALYTICS=true',
```
Comment (jonahwilliams) [inline packages/flutter_tools/test/general.shard/ios/ios_device_start_nonprebuilt_test.dart:55]: We actually want the analytics now that each re-entrant build corresponds to a single flutter invocation

File: packages/flutter_tools/lib/src/ios/mac.dart [hunk 3]
```diff
@@ -188,90 +209,89 @@ Future<XcodeBuildResult> buildXcodeProject({
...
+    buildInfo: buildInfo,
+    codesign: codesign,
+    buildForDevice: buildForDevice,
+    targetOverride: targetOverride,
+    deviceID: deviceID,
+  );
+  if (parsedProjectInfo == null) {
+    return XcodeBuildResult(success: false);
+  }
   final List<String> buildCommands = <String>[
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/ios/mac.dart:242]: I tried to collapse most of the logic for creating the xcodebuild command using if/for elements - it should be easier to read for the most part...

File: packages/flutter_tools/lib/src/ios/mac.dart [hunk 4]
```diff
@@ -395,49 +409,47 @@ Future<XcodeBuildResult> buildXcodeProject({
...
-    if (hasWatchCompanion && !buildForDevice) {
-      globals.printTrace('Replacing iphoneos with iphonesimulator in TARGET_BUILD_DIR.');
-      targetBuildDir = targetBuildDir.replaceFirst('iphoneos', 'iphonesimulator');
-    }
-    final String expectedOutputDirectory = globals.fs.path.join(
-      targetBuildDir,
-      buildSettings['WRAPPER_NAME'],
-    );
+  }
+  // If the app contains a watch companion target, the sdk argument of xcodebuild has to be omitted.
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/ios/mac.dart:413]: Logic is the same, just un-nested the else on old L398

**Review by jmagman — APPROVED** [review]
LGTM

**Comment by jonahwilliams** [comment]
Thanks!

**Pull request merged**